File: krass_jass/scoring.py
"""Round scoring.

Trick points, the last-trick bonus and the match bonus, all scaled by the contract
multiplier. Weis and Stöck are added by the caller (they are announced during play, not
derived from the tricks) and are scaled by the same multiplier.
"""

from __future__ import annotations

from dataclasses import dataclass

from .rules import Contract, RulesConfig

NUM_TEAMS = 2
TRICKS_PER_ROUND = 9


def team_of(seat: int) -> int:
    """Seats 0/2 are team 0, seats 1/3 are team 1."""
    return seat & 1


@dataclass(frozen=True)
class RoundScore:
    """Everything that went into one round's points, kept separate for the trace."""

    trick_points: tuple[int, int]
    last_trick: tuple[int, int]
    match: tuple[int, int]
    weis: tuple[int, int]
    stoeck: tuple[int, int]
    multiplier: int

    @property
    def total(self) -> tuple[int, int]:
        return tuple(
            (self.trick_points[t] + self.last_trick[t] + self.match[t] + self.weis[t] + self.stoeck[t])
            * self.multiplier
            for t in range(NUM_TEAMS)
        )


def score_round(
    trick_points: tuple[int, int],
    tricks_won: tuple[int, int],
    last_trick_winner: int,
    contract: Contract,
    cfg: RulesConfig,
    weis: tuple[int, int] = (0, 0),
    stoeck: tuple[int, int] = (0, 0),
) -> RoundScore:
    """Assemble one round's score.

    ``trick_points`` and ``tricks_won`` are per *team*; ``last_trick_winner`` is a seat.
    """
    last = [0, 0]
    last[team_of(last_trick_winner)] = cfg.last_trick_bonus

    match = [0, 0]
    if cfg.match_bonus:
        for t in range(NUM_TEAMS):
            if tricks_won[t] == TRICKS_PER_ROUND:
                match[t] = cfg.match_bonus

    return RoundScore(
        trick_points=tuple(trick_points),
        last_trick=tuple(last),
        match=tuple(match),
        weis=tuple(weis) if cfg.weis_enabled else (0, 0),
        stoeck=tuple(stoeck) if cfg.stoeck_enabled else (0, 0),
        multiplier=cfg.multiplier(contract),
    )

File: rust/src/scoring.rs
//! Round scoring and the Stöck-Weis-Stich claim ordering. Twin of `krass_jass/scoring.py`
//! and the claim sequence in `krass_jass/game.py`.

use crate::config::{team_of, Rules, CLAIM_STICH, CLAIM_STOECK, CLAIM_WEIS};

pub const TRICKS_PER_ROUND: usize = 9;

#[derive(Clone, Copy, Debug, Default)]
pub struct RoundScore {
    pub trick_points: [i32; 2],
    pub last_trick: [i32; 2],
    pub match_bonus: [i32; 2],
    pub weis: [i32; 2],
    pub stoeck: [i32; 2],
    pub multiplier: i32,
}

impl RoundScore {
    pub fn total(&self) -> [i32; 2] {
        let mut out = [0i32; 2];
        for t in 0..2 {
            out[t] = (self.trick_points[t]
                + self.last_trick[t]
                + self.match_bonus[t]
                + self.weis[t]
                + self.stoeck[t])
                * self.multiplier;
        }
        out
    }
}

pub fn score_round(
    trick_points: [i32; 2],
    tricks_won: [usize; 2],
    last_trick_winner: usize,
    contract: usize,
    rules: &Rules,
    weis: [i32; 2],
    stoeck: [i32; 2],
) -> RoundScore {
    let mut last = [0i32; 2];
    last[team_of(last_trick_winner)] = rules.last_trick_bonus;

    let mut match_bonus = [0i32; 2];
    if rules.match_bonus != 0 {
        for t in 0..2 {
            if tricks_won[t] == TRICKS_PER_ROUND {
                match_bonus[t] = rules.match_bonus;
            }
        }
    }

    RoundScore {
        trick_points,
        last_trick: last,
        match_bonus,
        weis: if rules.weis_enabled { weis } else { [0, 0] },
        stoeck: if rules.stoeck_enabled { stoeck } else { [0, 0] },
        multiplier: rules.multiplier(contract),
    }
}

/// Points in the order they are claimed: **Stöck, Weis, Stich**.
///
/// Only matters when both teams would cross the target in the same round — then whoever gets
/// there first in this order wins, whatever the final totals say. Counting the round in one
/// go produces the right totals and can name the wrong winner.
///
/// `trick_results` is `(winning seat, card points)` per trick, in the order taken.
pub fn claim_sequence(
    score: &RoundScore,
    trick_results: &[(usize, i32)],
    rules: &Rules,
) -> Vec<(usize, i32)> {
    let m = score.multiplier;
    let mut stoeck = Vec::new();
    let mut weis = Vec::new();
    let mut stich = Vec::new();

    for team in 0..2 {
        if score.stoeck[team] != 0 {
            stoeck.push((team, score.stoeck[team] * m));
        }
        if score.weis[team] != 0 {
            weis.push((team, score.weis[team] * m));
        }
    }

    let last = trick_results.len().saturating_sub(1);
    for (index, &(winner, points)) in trick_results.iter().enumerate() {
        let mut p = points;
        if index == last {
            p += rules.last_trick_bonus;
        }
        if p != 0 {
            stich.push((team_of(winner), p * m));
        }
    }
    for team in 0..2 {
        if score.match_bonus[team] != 0 {
            stich.push((team, score.match_bonus[team] * m));
        }
    }

    let mut out = Vec::new();
    for key in rules.claim_order {
        match key {
            CLAIM_STOECK => out.extend(stoeck.iter().copied()),
            CLAIM_WEIS => out.extend(weis.iter().copied()),
            CLAIM_STICH => out.extend(stich.iter().copied()),
            _ => {}
        }
    }
    out
}
